src/sinks/amqp.rs

```diff
@@ -0,0 +1,615 @@
+                            }
+                        }
+                        Err(error) => error!(message = "Amqp error.", %error),
+                    };
+
+                    this.pending_acks.insert(seqno);
+
+                    let mut num_to_ack = 0;
+                    while this.pending_acks.remove(&this.seq_tail) {
+                        num_to_ack += 1;
+                        this.seq_tail += 1
+                    }
+                    this.acker.ack(num_to_ack);
+                }
+                Some((_, Err(Canceled))) => {
+                    error!(message = "Request canceled.");
+                    return Poll::Ready(Err(()));
+                }
+                None => break,
+            }
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.poll_flush(cx)
+    }
+}
+
+async fn healthcheck(_config: AmqpSinkConfig, channel: Arc<lapin::Channel>) -> crate::Result<()> {
+    trace!("Healthcheck started.");
+
+    if !channel.status().connected() {
+        warn!("Channel is not connected");
+        return Err(Box::new(std::io::Error::new(std::io::ErrorKind::BrokenPipe, "Not Connected")));
+    }
+
+    trace!("Healthcheck completed.");
+    Ok(())
+}
+
+fn encode_event(
+    mut event: Event,
+    key_field: &Option<String>,
+    encoding: &EncodingConfig<Encoding>,
+) -> (String, Vec<u8>) {
+    let key = key_field
+        .as_ref()
+        .and_then(|f| match &event {
+            Event::Log(log) => log.get(f).map(|value| value.to_string_lossy()),
+            Event::Metric(metric) => metric
+                .tags()
+                .and_then(|tags| tags.get(f))
+                .map(|value| value.to_string()),
+        })
+        .unwrap_or_default();
+
+    encoding.apply_rules(&mut event);
+
+    let body = match event {
+        Event::Log(ref log) => match encoding.codec() {
+            Encoding::Json => serde_json::to_vec(log).unwrap(),
+            Encoding::Text => log
+                .get(log_schema().message_key())
+                .map(|v| v.as_bytes().to_vec())
+                .unwrap_or_default(),
+            Encoding::Protobuf => {
+                let event: proto::EventWrapper = event.into();
+                let mut payload = Vec::with_capacity(1_000);
+                event.encode(&mut payload).unwrap();
+                payload
+            },
+        },
+        Event::Metric(ref metric) => match encoding.codec() {
+            Encoding::Json => serde_json::to_vec(metric).unwrap(),
+            Encoding::Text => metric.to_string().into_bytes(),
+            Encoding::Protobuf => {
+                let event: proto::EventWrapper = event.into();
+                let mut payload = Vec::with_capacity(1_000);
+                event.encode(&mut payload).unwrap();
+                payload
+            },
+        },
+    };
```

Comment:
Could we swap the `unwrap`s for `expect`s here to self-document why we are ok with panicking? In this case, these encodings should never fail so I think it is fine to panic, but using `expect` will make that more clear.